emcc

```diff
@@ -808,7 +813,7 @@ try:
   # -c means do not link in gcc, and for us, the parallel is to not go all the way to JS, but stop at bitcode
   has_dash_c = '-c' in newargs
   if has_dash_c:
-    assert has_source_inputs or has_header_inputs, 'Must have source code or header inputs to use -c'
+    assert has_source_inputs or has_header_inputs or has_jsobj_inputs, 'Must have source code or header inputs to use -c'
```

Comment:
Good question, OK, I'll try and explain myself better.

What we'd like is the following sequence of commands:

    emcc -o lib1.c.o -c lib1.c              # Uses function in libjs.jso
    emar cr libexample.a lib1.c.o libjs.jso # Make lib containing jso file
    emcc -o main.c.o -c main.c              # Uses function in lib1.c
    emcc -o main.js main.c.o libexample.a

And indeed that works! Unfortunately, it's a real pain getting a build system like CMake to run those commands, because CMake is designed to compile things, and there's no "compile" step for the file `libjs.jso`. It doesn't expect to use raw files in static libraries.

So, if you were compiling with the following CMakeLists.txt file you'd generate a slightly different set of build commands:

    cmake_minimum_required(VERSION 2.8)
    project(testproj)
    add_library(example STATIC lib1.c libjs.jso)
    add_executable(main main.c)
    target_link_libraries(main example)

CMake will run the following commands:

    emcc -o lib1.c.o -c lib1.c            #     Same as before
    emcc -o libjs.jso.o -c libjs.jso      # *** "compile" the jso file (raw copy)
    emar cr libexample.a lib1.c.o libjs.jso.o # Include libjs.jso.o
    emcc -o main.c.o -c main.c            #     Same as before
    emcc -o main.js main.c.o libexample.a #     Same as before

So the hack in `***` is fairly simple: we allow build systems to run `emcc -o libjs.jso.o -c libjs.jso` (equivalent to `cp libjs.jso libjs.jso.o`), treating jso files just like all the other source files. Incremental builds pick up the dependencies beautifully.